unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+
+        for _, chunks in itertools.groupby(
+            sorted(
+                # filter out deleted chunks (page_status=0)
+                [chunk for chunk in self.chunks if chunk.page_status != 0x0],
+                key=lambda chunk: chunk.chunk_id,
+            )
+        ):
+            yield max(chunks, key=lambda chunk: chunk.serial)
+
+
+@attr.define
+class YAFFSFileVar:
+    file_size: int
+    stored_size: int
+    shrink_size: int
+    top_level: int
+
+
+@attr.define(kw_only=True)
+class YAFFS2Entry(YAFFSEntry):
+    chksum: int
+    yst_mode: int
+    yst_uid: int
+    yst_gid: int
+    yst_atime: int
+    yst_mtime: int
+    yst_ctime: int
+    equiv_id: int
+    yst_rdev: int
+    win_ctime: List[int]
+    win_mtime: List[int]
+    inband_shadowed_obj_id: int
+    inband_is_shrink: int
+    reserved: List[int]
+    shadows_obj: int
+    is_shrink: int
+    filehead: YAFFSFileVar
+    chunks: List[YAFFS2Chunk]
+
+    def get_chunks(self) -> Iterable[YAFFS2Chunk]:
+        """Return a filtered and ordered list of chunks."""
+        # The Yaffs2 sequence number is not the same as the Yaffs1 serial number!
+
+        # As each block is allocated, the file system's
+        # sequence number is incremented and each chunk in the block is marked with that
+        # sequence number. The sequence number thus provides a way of organising the log in
+        # chronological order.
+
+        # Since we're scanning backwards, the most recently written - and thus current - chunk
+        # matching an obj_id:chunk_id pair will be encountered first and all subsequent matching chunks must be obsolete and treated as deleted.
+
+        # note: there is no deletion marker in YAFFS2
+
+        for _, chunks in itertools.groupby(
+            sorted(self.chunks, key=lambda chunk: chunk.chunk_id)
+        ):
+            yield max(chunks, key=lambda chunk: chunk.seq_number)
+
+
+ROOT = YAFFS1Entry(
+    object_type=0,
+    object_id=-1,
+    parent_obj_id=-1,
+    sum_no_longer_used=0,
+    name=".",
+    alias="",
+    file_size=0,
+    start_offset=0,
+    chunks=[],
+)
+
+
+SPARE_START_BIG_ENDIAN_ECC = b"\x00\x00\x10\x00"
+SPARE_START_BIG_ENDIAN_NO_ECC = b"\xFF\xFF\x00\x00\x10\x00"
+SPARE_START_LITTLE_ENDIAN_ECC = b"\x00\x10\x00\x00"
+SPARE_START_LITTLE_ENDIAN_NO_ECC = b"\xFF\xFF\x00\x10\x00\x00"
+PAGE_START_LITTLE_ENDIAN = b"\x01\x00\x00\x00\x01\x00\x00\x00\xff\xff"
+PAGE_START_BIG_ENDIAN = b"\x00\x00\x00\x01\x00\x00\x00\x01\xff\xff"
+
+# 00 00 c0 ff ff ff 01 00  9a aa a7 a4 c1 59 aa ab
+# 00 00 c0 ff ff ff 01 01  03 ff 0f 00 c1 59 aa ab
+# 00 00 c0 ff ff ff 02 01  0f ff f3 0c c1 fc ff 03
+
+# 00 00 03 ff ff ff 00 00  95 aa a7 40 3f 56 aa ab
+# 00 00 03 ff ff ff 00 40  0c ff 0f 40 83 56 aa ab
+# 00 00 03 ff ff ff 00 40  00 ff f3 80 bf fc ff 03
+
+
+class YAFFSParser:
+    def __init__(self, file: File, config: Optional[YAFFSConfig] = None):
+        self.file_entries = Tree()
+        self.file_entries.create_node(ROOT, 1)
```

Comment:
isn't it v1 specific? Also using a global singleton wouldn't be a problem when multiple yffs identified in one unblob run?

Reply:
We need it for version 2 since they don't store an empty root node to gain space (as opposed to version 1). So we create this root node explicitly.

This has been moved to the specific `YAFFS2Parser`. I'm now using a local variable rather than a global singleton so we should be safe in terms of sharing objects.